Amqp

manuelstein opened the pull request:
# Pull request

## Description

Adding source and sink for AMQP using lapin.

## Related

* Related [docs PR](https://github.com/tremor-rs/tremor-www-docs/pull/139)

## Checklist

* [X] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [X] The code is tested
* [x] Use of unsafe code is reasoned about in a comment (N/A)
* [X] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [X] The performance impact of the change is measured (see below)

## Performance

adding an onramp/offramp wouldn't affect workflow processing, would it?

Licenser commented on src/sink/amqp.rs:
```diff
@@ -0,0 +1,268 @@
+pub struct Config {
+    pub amqp_addr: String,
+    routing_key: String,
+    exchange: String,  // "", by default
```
Adding
```rust
    #[serde(default = "Default::default")]`
```
to this would make it optional and default to `""` when nothing is provided

manuelstein replied:
that's a great idea, I checked that both routing_key and exchange can be blank in amqp, which is the default and in fanout mode even is of no use (so blank string it is).

Licenser commented on src/sink/amqp.rs:
```diff
@@ -0,0 +1,268 @@
+}
+
+impl Amqp {
+    fn handle_channel(&mut self) -> Result<Option<&Channel>> {
```
this this could in result of `channel` being async also become async

Licenser commented on src/sink/amqp.rs:
```diff
@@ -0,0 +1,268 @@
+                    */
+                    // prepare message headers
+                    let properties = BasicProperties::default();
+                    /*
```
Hades would be really nice, but not a blocker, lets throw in a ticket to add them later  if you're OK with that

Licenser replied:
I've no idea what I mean with this so I'll close it :joy:

manuelstein replied:
Headers postponed for now due to lack of time, sry.

Licenser commented on src/source/amqp.rs:
```diff
@@ -0,0 +1,236 @@
+    queue_name: String,
+    queue_options: QueueDeclareOptions,
+    routing_key: String,
+    exchange: String,  // "", by default
+    #[serde(default = "Default::default")]
+    pub close_on_done: bool,
+    #[serde(default = "Default::default")]
+    pub sleep_on_done: u64,
```
see the nats src, I think the same applies to this config

manuelstein replied:
as above, removed these two residuals

Licenser commented on src/source/amqp.rs:
```diff
@@ -0,0 +1,236 @@
+        if self.consumer.is_none() {
+            return Ok(SourceReply::StateChange(SourceState::Disconnected))
+        }
+        if let Some(delivery) = self.consumer.as_mut().unwrap().next().await {
```
we don't allow unwrap or expect, instead we would prefer errors bubbling up in this cases.

The reasoning here is that if a single source or sink has an issue we'd not want to panic the entire application and potentially take down unrelated and pipelines that themselves are unaffected. This gets more important since we allow loading configuration at run time, and someone making an API call could, if we'd allow unwrap/expect/panic/etc. then crash the application with providing a 'bad' configuration

manuelstein replied:
Resolved it into a match tree

Licenser reviewed (changes requested):
Heya,

thousand thanks for this :D it looks really good there are just a few small ponish things left, I mark them inline to make it easier to track :), if you need any help with them feel free to ping on discord.

manuelstein commented on src/sink/amqp.rs:
```diff
@@ -0,0 +1,268 @@
+    #[serde(default = "Default::default")]
+    pub close_on_done: bool,
+    #[serde(default = "Default::default")]
+    pub sleep_on_done: u64,
```
thanks, it's some copy'n'paste residual. removed

manuelstein commented on src/sink/amqp.rs:
```diff
@@ -0,0 +1,268 @@
+}
+
+impl Config {
+    fn channel(&self) -> Result<Channel> {
```
now, this was a tough one, but I got there eventually - even learning one or two things about Rust. I hope they stick :) thx!

Licenser commented:
This looks great :D you just got to rebase then it's good to go :rocket: thank you so much for this!

manuelstein commented:
Thank you for the code review!
Rebased onto main.
I'm still getting sanity errors on let _ variable use when the AMQP connection is terminated :-/
Used vscode auto-formatting and addressed as many clippies as I could, not sure what to do with the remaining.
Didn't update Cargo.lock, which seems to fail tests-linux.

manuelstein commented:
and thanks for the praise 😊❤️ at least it's a start

Licenser commented on src/source/amqp.rs:
```diff
@@ -0,0 +1,244 @@
+            Some(consumer) => {
+                match consumer.next().await {
+                    Some(delivery) => {
+                        let (_, delivery) = delivery.expect("error in consumer");
```
expect's can't be used but you can write it like this:
```
                        let (_, delivery) = delivery?;
```

Licenser commented:
Heya I want over the remaining issues and made three commits to go over them one by one, let me quickly post them and walk through them :)

[The `Cargo.lock`](https://github.com/tremor-rs/tremor-runtime/commit/e84e494e4831882a49379f6fe6c2b5c25887c9ee) this is easy every time cargo build is called the lock file is updated

[Safety checks](https://github.com/tremor-rs/tremor-runtime/commit/2c1e75319dc6dee57d4aee87bf19f3241e751bd2) there are two things here.

One is that the CI requires must-use results to be used and something is done with them, that's to avoid silent errors. the piny thing doesn't make it easy to figure out what to do but effectively we look at the results. The good news is this caught a missing `await` too so the code would not have been executed :).

The second one is the `expect` could be replaced with a `?` and we have nice errors now  :rocket: 

Last [Clippy](https://github.com/tremor-rs/tremor-runtime/commit/ccc480bdcd5706911fc9e260e79a44d23b5c43a8) our best friend and worst enemy :joy: the two things it fagged were:

A truncation, which makes sense, usually truncating a 128bit value into a 64 bit value can lead to problems. In this case, we know what it is used for so we allow it and add an explanation on my we allow it that way when in 100 years we stumble over that allow we know why it is there :D

The second thing was just the rust style Clippy enforces it prefers `something.map(some_function)` over `something.map(|v| some_function(v))` they do the same but Clippy keeps our code style consistent so it's nice, even so sometimes annoying ;)

Feel free to cherry-pick the commits or copy&paste the changes into your branch, I just wanted to split them out in a few different units so it's easy to explain what is going on behind the scenes, I hope it helps!

Other than that, all that's needed would be a git-rebase to prevent merge conflicts and :rocket: we're good to go :D

Licenser commented:
Wooooh :D nice

darach commented:
I love returning from time off work and finding a wonderful new contribution to tremor. Great work @manuelstein and i'm looking forward to giving this a whirl!

darach reviewed (approved):
LGTM 🎸
